demos/python_demos/face_recognition_demo.py

```diff
@@ -0,0 +1,1037 @@
+
+    def start_async(self, frame, rois):
+        inputs = self.preprocess(frame, rois)
+        for input in inputs:
+            self.enqueue(input)
+
+    def get_landmarks(self):
+        outputs = self.get_outputs()
+        results = [ self.Result(out[self.output_blob].reshape((-1, 2))) \
+                    for out in outputs ]
+        return results
+
+    def _adjust_color(self, frame):
+        assert len(frame.shape) == 4 and frame.shape[1] == 3, \
+            "Expected input of shape [1, 3, h, w] in BGR format"
+        image = frame[0]
+        return np.array([[image[2], image[1], image[0]]]) # convert BGR to RGB
+
+
+class FacesDatabase:
+    IMAGE_EXTENSIONS = ['.jpg', '.png']
+
+    class Identity:
+        def __init__(self, label, descriptor):
+            self.label = label
+            self.descriptor = descriptor
+
+    class ROI:
+        def __init__(self, x, y, w, h):
+            self.data = np.array((x, y, w, h))
+            self.position = self.data[0:2]
+            self.size = self.data[2:4]
+
+    def __init__(self, path, face_identifier, landmarks_detector):
+        path = osp.abspath(path)
+        paths = []
+        if osp.isdir(path):
+            ext = self.IMAGE_EXTENSIONS
+            paths = [osp.join(path, f) for f in os.listdir(path) \
+                      if f.endswith(ext[0]) or f.endswith(ext[1])]
+        else:
+            raise Exception("Wrong face images database path. Expected path to" \
+                            "a directory containing '%s' files, got '%s'" % \
+                            (self.IMAGE_EXTENSIONS, path))
+
+        self.database = []
+        for path in paths:
+            label = osp.splitext(osp.basename(path))[0]
+            image = cv2.imread(path)
+
+            n, c, h, w = face_identifier.get_input_shape()
+            image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
+            image = cv2.resize(image, (w, h))
+            image = image.transpose((2, 0, 1)) # HWC to CHW
+            image = image.reshape((n, c, h, w))
+
+            rois = [ self.ROI(0, 0, w, h) ]
+            landmarks_detector.start_async(image, rois)
+            landmarks = landmarks_detector.get_landmarks()
+
+            face_identifier.start_async(image, rois, landmarks)
+            descriptor = face_identifier.get_descriptors()[0]
+            self.database.append(self.Identity(label, descriptor))
+
+    def match_faces(self, descriptors):
+        database = self
+        distances = np.empty((len(descriptors), len(database)))
+        for i, desc in enumerate(descriptors):
+            for j, identity in enumerate(database):
+                distances[i][j] = cosine(desc, identity.descriptor) * 0.5
+
+        # Find best assignments, prevent repeats (assuming faces can not repeat)
+        _, assignments = linear_sum_assignment(distances)
+        matches = []
+        for i in range(len(descriptors)):
+            id = assignments[i]
+            distance = distances[i, id]
```

Comment:
When for some `i` distance is close to 1, the `i`th face is unlikely to be a person from the database.
IMHO here should be thresholding: something like if distance > 0.8 -> person is unknown.